### PR TITLE
builder: implement basic delta shard functioanlity

### PR DESCRIPTION
**Pull request by ggilmore**
This PR adds the following delta shard functionality to the builder:

- if the `IsDelta` builder.Option field is `true` (indicating that this build process is generating a delta shard), the builder now does the following:
	- properly sets the shard numbering for delta shards: given the preexisting shards `shard-0`, `shard-1`, `shard-2` - the builder knows to start the naming scheme for the new delta shards with `shard-3`. This allows our `FindAllShards` logic to discover the new delta shards along side the old ones. 
	- skips deleting preexisting shards: Unlike normal shards, delta shards rely on stacking changes on top of older ones. So, we need to keep the preexisting ones around. The previous behavior (deleting all older shards) is maintained for normal builds. 
	- updates the FileTombstone data in older shards: The `ChangeOrModifiedFiles` field in the builder options lets the builder know what files need to marked as tombstoned in the preexisting shards. 
	- updates the branch version information in all older shards: Our `findShard`/ `IndexState()` functionality implicitly assumes that the branch version information is _consistent_ across all shards. Previously, this was always true since we deleted all the older shards for every build. With delta shards, we're able to maintain this property by editing the branch metadata in all the older shards to match the information in the current build. 
		- **Note**: we can only do this if the _set_ of branch names hasn't changed between builds (only the commits they point to have changed). If the set of names has changed, updating this information in the older shards won't work (it'll cause panics in searcher). 

- `IndexState()` - if IsDelta is set to true (acting here a feature flag), it now returns more granular information that allows us to tell if the set of branch names or commits specifically have changed. In the future, this will allow callers (like `zoekt-git-index` to know if they can get away with only a delta shard build (only the commits have changed) or if they need to fall back to doing a full normal build (if the set of branch names has changed - or any other relevant field). 

---

Another note: This PR doesn't include robust behavior for being able to recover if the builder fails in the middle of a delta shard build. Including in this PR doesn't seem necessary since everything is behind the `IsDelta` field (acting as a feature flag). For more context, see the "dirty file" discussion at the bottom of https://sourcegraph.com/notebooks/Tm90ZWJvb2s6MjIx

### CODE REVIEW

File: build/builder.go [hunk 1]
```diff
@@ -350,8 +363,28 @@ func (o *Options) IndexState() (IndexState, string) {
...
+		if !cmp.Equal(repo.Branches, o.RepositoryDescription.Branches, sortBranchesOption) {
+			return IndexStateBranchVersion, fn
+		}
+	} else {
+		if !reflect.DeepEqual(repo.Branches, o.RepositoryDescription.Branches) {
+			return IndexStateContent, fn
+		}
 	}
 
 	// We can mutate repo since it lives in the scope of this function call.
```
Comment (ggilmore) [inline build/builder.go:386]: cc @stefanhengl I decided to not add the more granular branch indexstate information to MergeMutable since the only use of this is for compound shards. Since we're punting any work on that for now - no delta shard should end up in a compound shard (and the coarse `branches` check in merge mutable ensures that the function would error anyway), it doesn't seem necessary here

File: tombstones.go [hunk 2]
```diff
@@ -48,24 +48,41 @@ func setTombstone(shardPath string, repoID uint32, tombstone bool) error {
...
+// The filename name is generated by taking pattern and adding a random string to the end.
+// If pattern includes a "*", the random string replaces the last "*".
+// If dir is the empty string, CreateTemp uses the default directory for temporary files, as returned by os.TempDir.
+// It is the caller's responsibility to remove the file when it is no longer needed.
+// TODO: Should we stick this in a util package?
+// TODO: Maybe we should give this the path of a shard, and it returns a 3 tuple of (finalFile, tempFile, error)?
+//       If we do this, we can have it automatically figure out whether it needs to write out []zoekt.Repository or just
+//       zoekt.Repository. This would result in duplicate calls to zoekt.ReadMetadataPath at some of the callsites though
+//       such as the one above.
+func JsonMarshalRepoMetaTemp(repositoryMetadata interface{}, dir, pattern string) (tempFile string, err error) {
```
Comment (ggilmore) [inline tombstones.go]: I'm not super happy with this function. We now have multiple places that need to write repository metadata (shard merging and the builder for delta shards), and there is some nuance here with the `chmod` logic that all all of these places need to maintain. I think a shared function is the answer to that, but something about the interface feels awkward to me.

File: build/builder.go [hunk 3]
```diff
@@ -539,29 +583,100 @@ func (b *Builder) Finish() error {
...
+					// the only error we'll get is a panic when we actually search the shard.
+					r.Branches = b.opts.RepositoryDescription.Branches
+
+					break
+				}
+			}
+
+			var updatedRepositories interface{}
+
+			if indexMetadata.IndexFormatVersion >= 17 {
```
Comment (ggilmore) [inline build/builder.go]: I feel like this check is a bit of a footgun (needing to remember to check the index version to know how to serialize the repository metadata file). We also need to do this in both the shard merger and now in the builder for delta shards. Perhaps we should write a shared utility for this?

File: tombstones.go (hunk 2 above)
Comment (ggilmore) [reply tombstones.go]: Perhaps combining this function with the version check information (https://github.com/sourcegraph/zoekt/pull/274/files#r813424230) would make this interface a bit deeper and more useful?

**Comment by ggilmore** [comment]
@stefanhengl PTAL?

File: build/builder.go [hunk 4]
```diff
@@ -539,29 +583,100 @@ func (b *Builder) Finish() error {
...
+					}
+
+					for _, f := range b.opts.ChangedOrRemovedFiles {
+						r.FileTombstones[f] = struct{}{}
+					}
+
+					// TODO: Should we fail early here if we detect that the set of branch names
+					// in the builder options differs from what's inside the repository? Otherwise,
+					// the only error we'll get is a panic when we actually search the shard.
+					r.Branches = b.opts.RepositoryDescription.Branches
```
Comment (ggilmore) [inline build/builder.go]: Thoughts on this? I think the caller needs to check about what kind of index they'll need to do with IndexState() anyway, but adding another check here and setting `buildError` could add another layer of defense.

File: tombstones.go (hunk 2 above)
Comment (stefanhengl) [reply tombstones.go]: the interface{} doesn't feel right. What do think about something like this?

```
JsonMarshalRepoMetaTemp(dir string, md *api.Repository) (tmpFile, dstFile string, err error)
```

I imagine this is only called during the build process (?), so we don't have to worry about compound shards.

File: build/builder.go (hunk 4 above)
Comment (stefanhengl) [reply build/builder.go]: I agree. An additional check here will probably never fire, but if it does we already know that something went wrong between indexserver, which apparently didn't detect a mistmatch, and builder.

File: build/builder.go [hunk 5]
```diff
@@ -539,29 +583,100 @@ func (b *Builder) Finish() error {
...
+		// existing shards
+		for _, shard := range oldShards {
+			repositories, indexMetadata, err := zoekt.ReadMetadataPathAlive(shard)
+			if err != nil {
+				b.buildError = err
+				continue
+			}
+
+			for _, r := range repositories {
+				if r.ID == b.opts.RepositoryDescription.ID {
```
Comment (stefanhengl) [inline build/builder.go]: we could flip this and avoid the indentation? WDYT?
```
if !(r.ID == b.opts.RepositoryDescription) {
    continue
}
```

File: build/builder.go [hunk 6]
```diff
@@ -539,29 +583,100 @@ func (b *Builder) Finish() error {
...
+			} else {
+				if len(repositories) == 0 {
+					b.buildError = fmt.Errorf("failed to update repository metadata for shard %q - shard contains no repositories", shard)
+					continue
+				}
+
+				updatedRepositories = repositories[0]
+			}
+
+			finalPath := shard + ".meta"
```
Comment (stefanhengl) [inline build/builder.go]: I think it would be better if `JsonMarshalRepoMetaTemp` returned tmpPath and finalPath. `JsonMarshalRepoMetaTemp` could take care of all the tmp naming so the caller doesn't need to know about the details like the pattern.

File: build/builder.go (hunk 3 above)
Comment (stefanhengl) [reply build/builder.go]: Can't we skip this entirely? The builder never builds compound shards. Merge doesn't even call it. We can savely assume that we only every write out v16 shards here, can't we?

**Review by stefanhengl — COMMENTED** [review]
First pass

File: build/e2e_test.go [hunk 7]
```diff
@@ -600,3 +603,233 @@ func TestEmptyContent(t *testing.T) {
...
+						o.IsDelta = true
+						o.ChangedOrRemovedFiles = []string{"foo.go"}
+					},
+					query:             "common",
+					expectedDocuments: []zoekt.Document{barAtMain},
+				},
+			},
+		},
+		{
+			name: "tombstones are cumulative",
```
Comment (stefanhengl) [inline build/e2e_test.go]: Is there a difference between this test at the first one? The only difference I see is that this test doesn't add new documents. Is that relevant to test that tombstones are cumulative?

File: build/builder.go [hunk 8]
```diff
@@ -539,29 +583,100 @@ func (b *Builder) Finish() error {
...
+			repositories, indexMetadata, err := zoekt.ReadMetadataPathAlive(shard)
+			if err != nil {
+				b.buildError = err
+				continue
+			}
+
+			for _, r := range repositories {
+				if r.ID == b.opts.RepositoryDescription.ID {
+
+					if len(b.opts.ChangedOrRemovedFiles) > 0 && r.FileTombstones == nil {
```
Comment (stefanhengl) [inline build/builder.go]: I think we only need to set tombstones if `len(b.opts.ChangeOrRemovedFiles)>0`. Maybe it would make sense to move this check to line 594 and we might be able to skip a lot of work if files were just added.

File: build/builder.go [hunk 9]
```diff
@@ -539,29 +583,100 @@ func (b *Builder) Finish() error {
...
+						r.FileTombstones = make(map[string]struct{})
```
Comment (stefanhengl) [inline build/builder.go]: In this case we can prealocate the map with len(b.opts.ChangeOrRemovedFiles)

File: build/e2e_test.go [hunk 10]
```diff
@@ -600,3 +603,233 @@ func TestEmptyContent(t *testing.T) {
 		t.Errorf("got %+v, want 1 repo.", result.Repos)
 	}
 }
+
+func TestDeltaShards(t *testing.T) {
```
Comment (stefanhengl) [inline build/e2e_test.go:607]: I like this test!

**Comment by stefanhengl** [comment]
Just thinking out loud: Maybe we should, additional to a feature flag, have something like an allowlist for repos. In the first iteraton we could just put the mega-/gigarepo.

File: tombstones.go (hunk 2 above)
Comment (ggilmore) [reply tombstones.go]: I wanted to share this logic between zoekt-merge and the builder, which would require supporting both the normal and compound shard use-case. Maybe it's not worth it, however?

File: build/builder.go (hunk 3 above)
Comment (ggilmore) [reply build/builder.go]: I added this check because I was seeing an error if one of the pre-exisitng shards was a compound shard. Perhaps it'd be better just make it fail with an error like ("can't create a delta shard for a repository inside a compound shard" for now though).

**Comment by ggilmore** [comment]
> Just thinking out loud: Maybe we should, additional to a feature flag, have something like an allowlist for repos. In the first iteraton we could just put the mega-/gigarepo.

@stefanhengl 

I think that makes sense, but I don't think it needs to live in this PR. I think this logic is better suited for something like `zoket-git-index` since that component is the one that needs to decide whether it can do a delta shard build depending on the current state anyway.

File: build/e2e_test.go (hunk 7 above)
Comment (ggilmore) [reply build/e2e_test.go]: You're correct. I made some changes to the first test that makes this test redundant. I'll remove it.

File: build/builder.go (hunk 8 above)
Comment (ggilmore) [reply build/builder.go]: We still need to update the version metadata for all the older shards even if files were just added.

Comment (ggilmore) [reply build/builder.go]: This situation is actually tested in https://github.com/sourcegraph/zoekt/pull/274/files#diff-620acbfcb363a777e9bca5369361398fbc299ddaa61a1670402550375c914c59R426

File: build/builder.go (hunk 4 above)
Comment (ggilmore) [reply build/builder.go]: Done, and I also added a test for this here: https://github.com/sourcegraph/zoekt/blob/4d5062705c456709eae412252591a18d281a85aa/build/builder_test.go#L427-L459. 

Let me know if you think the test itself is worth it

File: build/builder.go (hunk 3 above)
Comment (ggilmore) [reply build/builder.go]: I added the error here: https://github.com/sourcegraph/zoekt/pull/274/commits/91525b3fe8c7ba910684b2759aef4af1ce91a74e . I was able to get rid of the for loop logic as a side effect.

File: tombstones.go (hunk 2 above)
Comment (ggilmore) [reply tombstones.go]: I changed the signature of the function [here](https://github.com/sourcegraph/zoekt/pull/274/commits/8d3e8f853c7d02add4724ddff32a2d3adba238ac): https://github.com/sourcegraph/zoekt/blob/8d3e8f853c7d02add4724ddff32a2d3adba238ac/tombstones.go#L63-L70

If you still think it's not worth support both the simple and compound shard usecase with this function (the reason why the signature includes `interface{}`), I can just duplicate the logic and create a separate function for just the builder's use-case.

File: build/builder.go (hunk 3 above)
Comment (stefanhengl) [reply build/builder.go]: Maybe we could we check in IndexState() and only trigger a delta build if the repo was not inside a compound shard?

File: build/builder.go (hunk 4 above)
Comment (stefanhengl) [reply build/builder.go]: The test is nice. I would keep it.

File: build/builder.go (hunk 8 above)
Comment (stefanhengl) [reply build/builder.go]: Oh yes, I forgot about the version metadata 🤦 .

File: build/builder.go [hunk 11]
```diff
@@ -539,29 +579,100 @@ func (b *Builder) Finish() error {
...
+		// Delta shard builds need to update FileTombstone and branch commit information for all
+		// existing shards
+		for _, shard := range oldShards {
+			repositories, indexMetadata, err := zoekt.ReadMetadataPathAlive(shard)
+			if err != nil {
+				b.buildError = err
+				continue
+			}
+
+			if indexMetadata.IndexFormatVersion > 16 {
```
Comment (stefanhengl) [inline build/builder.go]:
```suggestion
			if len(repositories) > 1 {
```

**Review by stefanhengl — APPROVED** [review]

File: tombstones.go (hunk 2 above)
Comment (stefanhengl) [reply tombstones.go]: I am ok with leaving it like this.
